### PR TITLE
Feature/feature report comment

### LINKED ISSUE #76: Delete Message deletes the user's first message on DB but not the specified date message

### PR DESCRIPTION
**Pull request by marclll**
close #76

### CODE REVIEW

File: src/main/java/org/pesmypetcare/webservice/thirdpartyservices/adapters/firestore/FirestoreCollectionAdapter.java [hunk 1]
```diff
@@ -242,7 +242,7 @@ private void checkArguments(Object[] fieldsAndValues) {
             if (fieldPath == null || fieldPath.toString().isEmpty()) {
                 throw new IllegalArgumentException(INVALID_FIELD_MESSAGE);
             }
-            query.whereEqualTo(fieldPath, value);
+            query = query.whereEqualTo(fieldPath, value);
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/thirdpartyservices/adapters/firestore/FirestoreCollectionAdapter.java:245]: Codacy Issue found: [Assignment of parameter 'query' is not allowed.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 14
           

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/error/InvalidOperationException.java  1
- src/main/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoImpl.java  3
- src/main/java/org/pesmypetcare/webservice/error/RestExceptionHandler.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182&bid=18448969)

File: src/test/java/org/pesmypetcare/webservice/controller/communitymanager/ForumRestControllerTest.java [hunk 2]
```diff
@@ -131,10 +133,29 @@ public void postMessage() throws Exception {
 
     @Test
     public void deleteMessage() throws Exception {
-        willDoNothing().given(service).createForum(anyString(), any(ForumEntity.class));
+        willDoNothing().given(service).deleteMessage(anyString(), anyString(), anyString(), anyString(), anyString());
+        mockMvc.perform(
+            delete(BASE_URL + parentGroup + "/" + forumName).header("token", myToken)
+                .param("creator", creator).param("date", creationDate)).andExpect(status().isNoContent());
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/controller/communitymanager/ForumRestControllerTest.java:139]: Codacy Issue found: [The String "creator" appears 5 times in the file.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182)

File: src/test/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoTest.java [hunk 3]
```diff
@@ -396,6 +404,45 @@ public void deleteMessage() throws DatabaseAccessException, DocumentException {
                     verify(batch).delete(same(documentReference));
                 }
 
+                @Test
+                public void reportMessage() throws DatabaseAccessException, DocumentException, InvalidOperationException {
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoTest.java:408]: Codacy Issue found: [Line is longer than 120 characters (found 122).](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182)

File: src/main/java/org/pesmypetcare/webservice/thirdpartyservices/adapters/firestore/FirestoreCollectionAdapter.java [hunk 4]
```diff
@@ -288,7 +288,7 @@ private void checkArguments(Object[] fieldsAndValues) {
             if (fieldPath == null || fieldPath.toString().isEmpty()) {
                 throw new IllegalArgumentException(INVALID_FIELD_MESSAGE);
             }
-            query.whereArrayContains(fieldPath, value);
+            query = query.whereArrayContains(fieldPath, value);
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/thirdpartyservices/adapters/firestore/FirestoreCollectionAdapter.java:291]: Codacy Issue found: [Assignment of parameter 'query' is not allowed.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182)

File: src/main/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoImpl.java [hunk 5]
```diff
@@ -174,6 +180,35 @@ public void deleteMessage(String parentGroup, String forumName, String creator,
         documentAdapter.commitBatch(batch);
     }
 
+    @Override
+    public void reportMessage(String parentGroup, String forumName, String creator, String reporter, String date)
+        throws DatabaseAccessException, DocumentException, InvalidOperationException {
+        if (creator.equals(reporter)) {
+            throw new InvalidOperationException("409", "A message creator can't report its message");
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoImpl.java:187]: Codacy Issue found: [The String "409" appears 2 times in the file.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182)

File: src/main/java/org/pesmypetcare/webservice/service/communitymanager/ForumServiceImpl.java [hunk 6]
```diff
@@ -102,6 +104,27 @@ public void deleteMessage(String token, String parentGroup, String forumName, St
         }
     }
 
+    @Override
+    public void reportMessage(String token, String parentGroup, String forumName, String creator, String reporter,
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/service/communitymanager/ForumServiceImpl.java:108]: Codacy Issue found: [More than 5 parameters (found 6).](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182)

File: src/main/java/org/pesmypetcare/webservice/thirdpartyservices/adapters/firestore/FirestoreCollectionAdapter.java [hunk 7]
```diff
@@ -265,7 +265,7 @@ private void checkArguments(Object[] fieldsAndValues) {
             if (field == null || field.isEmpty()) {
                 throw new IllegalArgumentException(INVALID_FIELD_MESSAGE);
             }
-            query.whereArrayContains(field, value);
+            query = query.whereArrayContains(field, value);
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/thirdpartyservices/adapters/firestore/FirestoreCollectionAdapter.java:268]: Codacy Issue found: [Assignment of parameter 'query' is not allowed.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182)

File: src/main/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoImpl.java [hunk 8]
```diff
@@ -174,6 +180,35 @@ public void deleteMessage(String parentGroup, String forumName, String creator,
...
+        }
+        DocumentSnapshot messageSnapshot = getForumMessage(parentGroup, forumName, creator, date);
+        WriteBatch batch = documentAdapter.batch();
+        batch.update(messageSnapshot.getReference(), REPORTED_BY_FIELD, FieldValue.arrayUnion(reporter));
+        ArrayList<String> messages = (ArrayList<String>) messageSnapshot.get(REPORTED_BY_FIELD);
+        if (messages != null && messages.contains(reporter)) {
+            throw new InvalidOperationException("409", "This user already reported the message");
+        }
+        if (messages != null && messages.size() > COUNTER-1) {
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoImpl.java:196]: Codacy Issue found: ['-' is not followed by whitespace.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182)

File: src/main/java/org/pesmypetcare/webservice/service/communitymanager/ForumService.java [hunk 9]
```diff
@@ -104,6 +105,35 @@ void postMessage(String token, String parentGroup, String forumName, Message mes
...
+     * @param parentGroup The parent group name
+     * @param forumName The forum name
+     * @param creator The creator's username
+     * @param reporter The reporter's username
+     * @param date The message publication date
+     * @throws DatabaseAccessException When the retrieval is interrupted or the execution fails
+     * @throws DocumentException When either the group or forum do not exist
+     * @throws InvalidOperationException When the operation isn't allowed
+     */
+    void reportMessage(String token, String parentGroup, String forumName, String creator, String reporter, String date)
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/service/communitymanager/ForumService.java:121]: Codacy Issue found: [More than 5 parameters (found 6).](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182)

File: src/main/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoImpl.java [hunk 10]
```diff
@@ -18,12 +19,14 @@
...
 import org.pesmypetcare.webservice.error.DatabaseAccessException;
 import org.pesmypetcare.webservice.error.DocumentException;
+import org.pesmypetcare.webservice.error.InvalidOperationException;
 import org.pesmypetcare.webservice.thirdpartyservices.FirebaseFactory;
 import org.pesmypetcare.webservice.thirdpartyservices.adapters.firestore.FirestoreCollection;
 import org.pesmypetcare.webservice.thirdpartyservices.adapters.firestore.FirestoreDocument;
 import org.pesmypetcare.webservice.utilities.UTCLocalConverter;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.stereotype.Repository;
+import org.springframework.web.client.HttpClientErrorException;
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoImpl.java:29]: Codacy Issue found: [Unused import - org.springframework.web.client.HttpClientErrorException.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182)

File: src/main/java/org/pesmypetcare/webservice/thirdpartyservices/adapters/firestore/FirestoreCollectionAdapter.java [hunk 11]
```diff
@@ -219,7 +219,7 @@ private void checkArguments(Object[] fieldsAndValues) {
             if (field == null || field.isEmpty()) {
                 throw new IllegalArgumentException(INVALID_FIELD_MESSAGE);
             }
-            query.whereEqualTo(field, value);
+            query = query.whereEqualTo(field, value);
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/thirdpartyservices/adapters/firestore/FirestoreCollectionAdapter.java:222]: Codacy Issue found: [Assignment of parameter 'query' is not allowed.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182)

File: src/test/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoTest.java [hunk 12]
```diff
@@ -396,6 +404,45 @@ public void deleteMessage() throws DatabaseAccessException, DocumentException {
...
+                        .willReturn(query);
+                    given(documentSnapshot.getReference()).willReturn(documentReference);
+                    given(documentSnapshot.get(anyString())).willReturn(usernameList);
+                    given(batch.update(any(DocumentReference.class), anyString(), any())).willReturn(batch);
+
+                    dao.reportMessage(groupName, forumName, username, username2, date);
+                    verify(collectionAdapter)
+                        .getDocumentsWhereEqualTo(eq(Path.ofCollection(Collections.messages, groupId, forumId)),
+                            eq("creator"), eq(username), eq(publicationDate), eq(date));
+                    verify(batch).update(same(documentReference), eq("reportedBy"),
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoTest.java:421]: Codacy Issue found: [The String "reportedBy" appears 2 times in the file.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182)

File: src/test/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoTest.java [hunk 13]
```diff
@@ -396,6 +404,45 @@ public void deleteMessage() throws DatabaseAccessException, DocumentException {
...
+                        eq(FieldValue.arrayUnion(username2)));
+                    verify(batch).update(same(documentReference), eq("banned"),
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoTest.java:423]: Codacy Issue found: [The String "banned" appears 2 times in the file.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182)

File: src/main/java/org/pesmypetcare/webservice/controller/communitymanager/ForumRestController.java [hunk 14]
```diff
@@ -151,6 +152,45 @@ public void deleteMessage(@RequestHeader String token, @PathVariable String pare
...
+     * @param creator The creator's username
+     * @param reporter The reporter's username
+     * @param date The message publication date
+     * @throws DatabaseAccessException When the retrieval is interrupted or the execution fails
+     * @throws DocumentException When either the group or forum do not exist
+     * @throws InvalidOperationException When the operation isn't allowed
+     */
+    @PutMapping("/{parentGroup}/{forumName}/report_message")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void reportMessage(@RequestHeader String token, @PathVariable String parentGroup,
```
Comment (santidrj) [inline src/main/java/org/pesmypetcare/webservice/controller/communitymanager/ForumRestController.java:170]: Codacy Issue found: [More than 5 parameters (found 6).](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182)

**Review by oricat8 — DISMISSED** [review]

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/error/InvalidOperationException.java  1
- src/main/java/org/pesmypetcare/webservice/error/RestExceptionHandler.java  1
- src/main/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoImpl.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182&bid=18448969)

File: src/test/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoTest.java [hunk 15]
```diff
@@ -410,7 +458,8 @@ public void addUserToLikedByOfMessage() throws DatabaseAccessException, Document
                     verify(collectionAdapter)
                         .getDocumentsWhereEqualTo(eq(Path.ofCollection(Collections.messages, groupId, forumId)),
                             eq("creator"), eq(username), eq(publicationDate), eq(date));
-                    verify(batch).update(same(documentReference), eq("likedBy"), eq(FieldValue.arrayUnion(username)));
+                    verify(batch).update(same(documentReference), eq("likedBy"),
```
Comment (santidrj) [inline src/test/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoTest.java:461]: Codacy Issue found: [The String "likedBy" appears 2 times in the file.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Coverage decreased per file
===========================
- src/main/java/org/pesmypetcare/webservice/service/communitymanager/ForumServiceImpl.java  -4
         

Complexity increasing per file
==============================
- src/main/java/org/pesmypetcare/webservice/error/InvalidOperationException.java  1
- src/main/java/org/pesmypetcare/webservice/dao/communitymanager/ForumDaoImpl.java  3
- src/main/java/org/pesmypetcare/webservice/error/RestExceptionHandler.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care-Webservice/pullRequest?prid=5633182&bid=18448969)

**Review by oricat8 — APPROVED** [review]

**Review by santidrj — APPROVED** [review]
:+1:

**Pull request merged**